feat(sliding sync): optionally persist/restore the `pos` in/from DB, and use that for encryption sync

bnjbvr opened the pull request:
This introduces a new `SlidingSyncBuilder` method, `restore_pos_from_database()`, that when enabled, will save the `pos` values into the database, and restore them just before running another request.

This will show useful in contexts where multiple processes can spawn a sliding sync, and we'd like continuity over the range of `pos` values that are sent, given a fixed `conn_id`. In particular, this will prevent spurious session expirations, when alternating between the sliding sync running in one process or the other. An example of that is the iOS setup, where we have the main app running as its own process, and a separate process running to decrypt notifications.

This can be reviewed commit by commit; the latest commit includes extensive tests, including some for session expiration (as well as some changes to maintain the invariant that `pos` is always the same in memory and in DB).

Hywan commented on crates/matrix-sdk/src/sliding_sync/mod.rs:
```diff
@@ -483,15 +486,24 @@ impl SlidingSync {
+                    request.extensions.to_device.since = restored_fields.to_device_token;
+                }
+                if self.inner.restore_pos_from_database {
+                    if request.pos != restored_fields.pos {
```
Did you do that so that every time we try to build a request, we read the value from the storage? So if someone else writes into the cache, we will use the latest version from the cache? Do you also bet that every time we get a response from the server, we save `SlidingSync` in the cache, so… we will just land back on tracks with the previous workflow:

* response received from the server
* save `SlidingSync` into the cache
* next request uses `pos` from the response
* send the request

vs.


* response received from the server
* save `SlidingSync` into the cache
* next request uses `pos` from the response
* if `pos` is different from what we have in the storage, let's use it in the request
* send the request

Is it the spirit?

bnjbvr replied:
I think so, I'm not so sure what you're asking here, sorry.

The new flow is:

- a response is received from the server
- we update the in-memory pos based on that response,
- *then* cache the pos into disk storage

When generating the next request:

- if configured to do so, read the pos from the storage
  - either it was us who write it, and then it'll be the same as the one we have in memory
  - or another process ran a sync in the background, and updated the pos in db, so we must use the value read from the db

In both cases, getting the value from the DB is the right thing to do.

Hywan commented on crates/matrix-sdk/src/sliding_sync/builder.rs:
```diff
@@ -40,6 +40,7 @@ pub struct SlidingSyncBuilder {
     rooms: BTreeMap<OwnedRoomId, SlidingSyncRoom>,
     poll_timeout: Duration,
     network_timeout: Duration,
+    restore_pos_from_database: bool,
```
I would maybe rename this to `shared_pos`, because that's the intention behind all this: `pos` is shared by all sliding sync instance with the same ID.

bnjbvr commented on crates/matrix-sdk/src/sliding_sync/builder.rs:
```diff
@@ -255,14 +274,12 @@ impl SlidingSyncBuilder {
 
             client,
             storage_key: self.storage_key,
+            restore_pos_from_database: self.restore_pos_from_database,
```
The sliding sync needs to be configured with this option as well, to decide when generating a request whether it reads the pos from its internal memory or DB. Think of the following use case: a sliding sync on main app has been interrupted, main app goes to background, NSE process runs a sync and saves new pos in the DB; then the main app when it goes back to the foreground needs to reload the pos from the DB.

bnjbvr commented:
Thanks for the initial review! I've addressed the first batch of comments as well as I could, and I've split the test commit into more meaningful logical commits, which should help with review.

bnjbvr commented:
I realized over the weekend this won't work, because the `pos` value would be stored in the _state_ store. On iOS, the state store used in the notification client (NSE process) is an in-memory state store, especially so it does not perform concurrent writes to the main state store database on disk (that would require locking, making sure we properly find and invalidate all state store caches etc.).

We need a better story here, so putting back into draft until we do.

Hywan commented on crates/matrix-sdk/src/sliding_sync/builder.rs:
```diff
@@ -220,15 +222,26 @@ impl SlidingSyncBuilder {
+    /// multi-process scenarios, to save it into some shared storage so that one
+    /// sliding sync instance running across two different processes can
+    /// continue with the same sync position it had before being stopped.
+    pub fn share_pos(mut self) -> Self {
```
Should we put this method behind `#[cfg(feature = "e2e-encryption")]`? Otherwise it will just have no effect.

bnjbvr replied:
Agreed for this one...

Hywan commented on crates/matrix-sdk/src/sliding_sync/mod.rs:
```diff
@@ -91,6 +91,10 @@ pub(super) struct SlidingSyncInner {
 
+    /// Should this sliding sync instance try to restore its sync position
+    /// from the database?
+    share_pos: bool,
```
Again, behind `cfg(feature = "e2e-encryption)`?

bnjbvr replied:
... but this one would make the code really harder to read (you'll see that to accomodate the above request, I'm already adding lots of cfg-guarded statements). i'm slightly in favor of keeping it to `false` for non-`e2e-encryption` builds. Thoughts?

Hywan replied:
OK. Let's add a comment to explain this maybe?

bnjbvr replied:
Sounds good!

Hywan reviewed (approved):
The patch is so clean. Best patch ever. Yummy. Delicious. Like a cake.

I wonder if the correct abstraction would not be a `RefCell`-ish type. Something that has interior mutability, and can be mutated by something else (like the storage) while we are owning it. That's really dangerous, I know.

Anyway, that's a terrible hack, and the plan is to remove this in a couple of weeks. Let's try.
